### ISSUE TITLE
New Feature Requests - leave comments here!

### ISSUE THREAD

**Issue by cliffordfajardo** [cliffordfajardo/cato#1]
If you have any new feature requests, please leave your comments below 👇

One of the goals I envision for Cato is for it to become a general purpose command launcher for your web browser; a browser extension that can handle automating **common manual tasks that people who use their web browsers do everyday.

# What's Not Possible (Platform Limitation)
- Opening the chrome developer tools/inspector programatically
  - ["We only allow explicit devtools opening."](https://bugs.chromium.org/p/chromium/issues/detail?id=112277)
- Restarting your browser programatically. 
   - `chrome://restart` can't be executed via code, it must be manually typed in the URL.

# Development Resources 

If you're a developer in interested learning about what's capable within the browser development ecosystem please take a look at the awesome resources:

- https://developer.mozilla.org/en-US/Add-ons/WebExtensions
- https://developer.chrome.com/extensions/api_index#dev_apis

**Comment by Naramsim** [comment]
Hi, thanks for the awesome extension 👍 

This is a list of features that could improve your project, in my opinion

- [ ] Go to a specific tab, by writing its name or the website in that tab
- [ ] A fancier style, would be awesome something like Discord (CMD+K) (screenshot below)
- [ ] Add new customizable search providers (Like Github search)
- [ ] When the input is 'theme', open the customize theme page
- [ ] Show commonly used commands at the top of the list
- [ ] Shortcut system as in [here](https://github.com/cliffordfajardo/cato/issues/1#issuecomment-332171886)


![cmdk](https://user-images.githubusercontent.com/8996268/30783758-b4b9b310-a148-11e7-9c6f-3d456512cd46.PNG)

**Comment by paradite** [comment]
I think a command to close duplicate tabs would be awesome.
There are some extensions that do this already, but they are not open source and extendable like this project, plus typing command is better than clicking mouse.

**Comment by cliffordfajardo** [comment]
@Naramsim - "Themes" for a command sounds like a good idea. For changing to a different tab, there's the "Change Tab" command already. 

![change-tabs](https://user-images.githubusercontent.com/6743796/30818711-2f9d69d6-a1d1-11e7-929a-2b2f079fe310.gif)

I had experimented  with this idea of just going _"to a specific tab, by writing its name or the website in that tab"_. It's definitely possible, I'm just worried that this will pollute the search results with too many commands initially. That's why I made the "Change Tab" command. 

You can change the keyboard shortcut of Cato to try and match discord in the extensions settings page: 
- https://www.howtogeek.com/127162/how-to-create-custom-keyboard-shortcuts-for-browser-actions-and-extensions-in-google-chrome/


@paradite - I can definitely create a command for closing duplicate tabs - that seems useful. In the next release, soon,  I'll get that out.

**Comment by Olian04** [comment]
I just started using this extensions, and I've go to say, I didn't know i needed this, but man did I need it!
Now for feature requests: 

__Commands:__
* [X] Add commands for showing sub- under the `chrome://` domain. Ex: `chrome://settings/help`, After closer inspection, the only sub-domain under the settings page that can be scrolled to from the settings page, is `/help`. 
* [X] Add command for shortening URLs. Ex: `Shorten URL` on this site would result in https://goo.gl/CBDRMd being copied to the clipboard.
* [ ] Add command for opening recently closed tabs.
* [ ] Add command for actively selecting the search command. As of now, the search command feels like a fallback option, "in case your query doesn't match a known command, lets search for it". It would be nice to see a command for actively choosing to search for a phrase. Ex: `search Bohemian Rhapsody ` would search for "Bohemian Rhapsody". Having a dedicated search command would allow for search term prediction / autocomplete.
* [ ] _(major)_ Add command for selecting DOM elements using CSS selectors. Ex: `Find input` would result in a list of all input fields on the page, moving the targeting to one of these would highlight the element on the page similar to how the devtools does it, then pressing enter would select the targeted element. If its an input it would open it for editing, if its a button it would press it, or if its a link it would open it, etc

__Behavior:__ 
* [ ] If you are on the "new page" page and a new page were to be opened as an effect of a command, the current page should be replaced. Ex: opening the settings page by using the `settings` command
* [ ] Pressing `Ctrl+Space` should show all available operations in alphabetical order. Aka pressing `Ctrl+Space` while in the command pallet should execute the `Show All Commands` command.
* [ ] Clicking on a result with the cursor should trigger the same code as navigating to it and pressing enter.

__Features:__
* [ ] Expose cross extension integration api. Possibly using the [Cross-extension messaging](https://developer.chrome.com/apps/messaging#external) feature of the chrome runtime. This would allow other installed extensions to hook into the command pallet and add their own commands.
* [ ] Allow for each command to specify in what context it should be executed. `background`, `injected`, `popup`, `options`, etc. As it stands every command is executed in the `popup` context.
* [ ] Allow for user defined commands. Maybe setup a page where users can associate their own JS code with a unique term that they then can use in the command pallet.  Ex: `dark-mode` could loop through the DOM and invert all colors.
* [ ] Allow for user specific aliases. Aka: `netflix` could be setup as an alias for `Change Tab netflix`. Or `c` could be used as a short hand for `Change Tab`. The important part is that the aliases should be easy to setup by the user.

**Comment by Naramsim** [comment]
@cliffordfajardo, I know there is such a command, but typing "change tab" slows me down 😄  Better to write directly the title of the tab I want to land, IMHO. Maybe you can disable this option by default, and let users decide in the configuration page if enabling it or not.

PS: I already changed my keys shortcut to CTRL+SHIFT+P, I suggested you the Discord shortcut so you can go check how they implemented the popup

Keep going man :), and thanks

**Comment by AlexisAnzieu** [comment]
Whaou, never thought i needed a tool like this. Thanks for it ! 
Right now, i've got some improvement suggestions :

- [ ] access to source code
- [ ] access to dev tools
- [ ] allow user to open their preferred website (as you do with amazon / gmail .... )
- [ ] cut sound from current page 

:1st_place_medal:

**Comment by Olian04** [comment]
@AlexisAnzieu What do you mean by `access to source code`? The source code for the extension is hosted right here, in this very repository. [Link just in case](https://github.com/cliffordfajardo/cato)

**Comment by Naramsim** [comment]
@Olian04 I think he means the command "View page source" (CTRL+U)

**Comment by Olian04** [comment]
@Naramsim ahh, alright. That's doable. Thanks for the clarifications :)

**Comment by Naramsim** [comment]
### Shortcut system

Would be awesome to add some shortcut i.e.:

yt <search> // search in youtube
gm <search> //search gmail
yt // go to youtube
ct <tab> // change tab
rb // remove from bookmarks current page

And all these shortcuts could be customizable by the user as suggested by @Olian04

**Comment by Arithmomaniac** [comment]
I'd like to second @Olian04 - user defined commands would be awesome.
I've been looking for a real [Firefox Ubiquity](https://wiki.mozilla.org/Labs/Ubiquity/Ubiquity_Source_Tip_Author_Tutorial) replacement for years, and [GHacks](https://www.ghacks.net/2017/09/25/cato-control-google-chrome-with-commands/) says you're halfway there 😃

**Comment by kenyarmosh** [comment]
+1 for just starting to type the name of a tab to switch to it. That's the main way I would hope to use it. Having to use "Change tab" really slows me down as noted by others above.

**Comment by Olian04** [comment]
@kenyarmosh @Naramsim Maybe there could be a second keybinding that opens the command palette with the command `Change Tab` already active?
Ex: `Crtl+K` opens a command pallet with the term `Change Tab` already in the query field.
![image](https://user-images.githubusercontent.com/5493406/30908522-171f37dc-a37e-11e7-9bb7-74abb4c8b08c.png)

**Comment by kenyarmosh** [comment]
@Olian04 That could work provided that we could still quickly start typing in the command palette to navigate the tabs.

**Comment by Olian04** [comment]
@AlexisAnzieu Your feature request for a command that `cut sound from current page` already exists :smile: The command is called `Mute/Unmute Tab`

As to your request for a command that opens the dev tools, by default chrome comes with a keybinding that opens the dev tools, on windows and linux that keybinding is `Ctrl+Shift+I` and on mac that keybinding is `Command+Option+I`. ~~Of course we could add a command for this as well.~~ But in the meantime theses shortcuts might be of help.

**Comment by Naramsim** [comment]
Yep, even F12 exists, and I consider really useful CTRL+C, which opens elements selector.

**Comment by dbautistav** [comment]
Feature request:
- [ ] Add custom **search** providers ([as in chromium](https://support.google.com/chrome/answer/95426?co=GENIE.Platform%3DDesktop&hl=en))

**Comment by Nolanus** [comment]
When interacting with tabs, I want the tab URL to be included in the search. For example I remember that I had a github tab somewhere. At the moment I can't search for "Github" to open it, as the page title does not contain that term.
In case of concerns about too many characters in some urls, resulting in a long list when searching for a tab, maybe just take the (sub)domain into account?!

**Comment by satomilumi** [comment]
delete history

**Comment by chiragbhansali** [comment]
- Basic Tab Navigation/Manipulation with keyboard shortcuts, similar to what Vimium can do. Like if I were to go one tab right or left, I can do that with some shortcut.
- Open URLs directly from the interface by typing them
- Create bookmarks for certain URLs in the Cato interface
- A feature which allows to check the meaning of a word

**Comment by PeytonHanel** [comment]
An option that allows the search feature to do a search in the current tab, rather than opening a new tab and doing it there.

**Comment by chiragbhansali** [comment]
A feature which allows to check the meaning of a word
